### PR TITLE
Feature/log mappings

### PR DESCRIPTION
**Pull request by MrHen**
Add custom hook mappings so we can more directly control what gets sent into the logging tools.

### CODE REVIEW

File: lib/autoBlock.js [hunk 1]
```diff
@@ -141,41 +215,62 @@ function finish (error, results, config) {
...
+  var withDeps = _.map(step.with, (value) => _.first(_.split(value, '.')))
 
-    // Build the async callback. This takes two forms: If no dependencies, no results paramater is included.
-    var asyncFunc
-    if (deps.length) {
-      asyncFunc = (results, cb) => runStep(results, cb, step, config)
-    } else {
-      asyncFunc = (cb) => runStep(null, cb, step, config)
+  var whenDeps = _.map(step.when, (value, key) => {
+    var dep = _.isString(key) ? key : value
```
Comment (justinmchase) [inline lib/autoBlock.js:235]: This is so `step.when` can be an object or an array? What can the values be if its an object?

File: lib/autoBlock.js [hunk 2]
```diff
@@ -141,41 +215,62 @@ function finish (error, results, config) {
...
 
-    asyncBlock[key] = deps
-  }, {})
+  deps = _.union(step.after, withDeps, whenDeps)
 
-  return autoBlock
+  // Build the async callback. This takes two forms: If no dependencies, no results paramater is included.
+  var asyncFunc
+  if (deps.length) {
+    asyncFunc = (results, cb) => runStep(results, cb, step, config)
```
Comment (justinmchase) [inline lib/autoBlock.js:249]: I'm not 100% sure but I think async will always take this function even if you have no dependencies.

File: lib/autoBlock.js [hunk 3]
```diff
@@ -292,6 +399,21 @@ function validateBlockSettings (step, stepName) {
...
 
+function validateHandlerSettings (step) {
+  var keys = _.keys(step)
+
+  var extra = _.reject(keys, (key) => {
+    return _.includes(validHandlerKeys, key)
+  })
+
+  if (extra.length) {
+    extra = _.map(extra, (key) => {
```
Comment (justinmchase) [inline lib/autoBlock.js]: Three things here:

1. This looks like a no-op, returning the same set as itself
2. short hand is better here:
```
_.map(extra, key => key)
```
3. lodash is not needed, map is already a function if you're in es6
```
extra.map(key => key)
```

File: lib/autoBlock.js [hunk 4]
```diff
@@ -292,6 +399,21 @@ function validateBlockSettings (step, stepName) {
...
+      return key
+    })
+    throw new Error('Invalid handler settings: ' + extra.join(', '))
```
Comment (justinmchase) [inline lib/autoBlock.js:413]: This function seems to throw if there are extra fields, but it doesn't seem to validate missing fields, such as `func` is that right? Or is `func` optional?

You could possibly do validation like this with joi in a cleaner way.

File: lib/autoBlock.js [hunk 5]
```diff
@@ -225,17 +320,29 @@ function runStepFunc (results, callback, step, config) {
...
+      step: {
+        name: step.name
+      },
+      data: config.data,
       result: !!result,
-      timer: (Date.now() - start) / 1000
-    }
-    config.onFinishStep(step.name, config.data, stepOutput)
+      results: results,
+      error: !!err,
```
Comment (justinmchase) [inline lib/autoBlock.js:344]: why not return the actual error object?

File: lib/autoBlock.js [hunk 6]
```diff
@@ -225,17 +320,29 @@ function runStepFunc (results, callback, step, config) {
...
+      timer: (Date.now() - start) / 1000,
```
Comment (justinmchase) [inline lib/autoBlock.js:345]: Why `/ 1000`? Why not leave it in raw form?

**Comment by justinmchase** [comment]
Seems like a minor version bump should be in there too.

**Review by justinmchase — APPROVED** [review]

File: lib/autoBlock.js (hunk 1 above)
Comment (MrHen) [reply lib/autoBlock.js:235]: `.when` is documented [here](https://github.com/TrackIF/auto-block/blob/505cb3a37aad49f5e049d2f32b8825d51d4ffbdb/README.md#when).

File: lib/autoBlock.js (hunk 2 above)
Comment (MrHen) [reply lib/autoBlock.js:249]: This behavior changed a lot when they were working on the latest version. I have a test covering the behavior so as long as they pass I'm happy.

File: lib/autoBlock.js (hunk 6 above)
Comment (MrHen) [reply lib/autoBlock.js:345]: Backwards compatibility for a dumb choice I made a while ago.

File: lib/autoBlock.js (hunk 5 above)
Comment (MrHen) [reply lib/autoBlock.js:344]: When we rev a major version number we can upgrade this to the full error object. For now, this functionality has been locked in.

(The short answer is that I was concerned about sensitive information appearing in logs and just mushed it into a boolean. But that shouldn't be the responsibility of _this_ code so it was the wrong thing to do.)

File: lib/autoBlock.js (hunk 3 above)
Comment (MrHen) [reply lib/autoBlock.js]: Uh, hm. That's definitely odd. Not sure why I had the extra step in there.

File: lib/autoBlock.js (hunk 4 above)
Comment (MrHen) [reply lib/autoBlock.js:413]: I'll see if I can add something real quick.

**Pull request merged**